### DISCUSSION TITLE
Week 4 homework2 POST 中文問題

### DISCUSSION THREAD

**Discussion by torai55** [Lidemy/forum#119] [程式導師實驗計畫] [answered]
來自用 https 套件寫的[作業二](https://github.com/Lidemy/mentor-program-5th-torai55/blob/master/homeworks/week4/hw2.js)。
嘗試使用 POST 新增中文書名會出錯：
![POST](https://user-images.githubusercontent.com/81942712/118509739-e74a0d00-b762-11eb-8ca2-2999b5bde34e.png)

看錯誤訊息好像是從 buffer 轉成字串的時候系統自動調用 `JSON.parse()`，但格式不對報錯。
試著用 [String Decoder](https://nodejs.org/api/string_decoder.html) 解析 buffer 也卡在 `JSON.parse`。
有發現輸入中文之後，回傳的 statusCode 是 500， content-type 變成 text/html 而不是 application/json。
猜測使用 `buffer.toString()` 時會先調用 `JSON.parse()` 轉成 JavaScript 物件之後才 `JSON.stringify()` 轉成字串，才因為回傳的 content-type 格式不對，在 parse 的步驟報錯。

---

![CLI](https://user-images.githubusercontent.com/81942712/118510300-6b03f980-b763-11eb-98e6-cad6c64690ca.png)
然後想順便問一下利用 curl POST 中文也會出現亂碼，該怎麼處理

**Comment by cwc329** [comment]
速解：
Content-length 帶的值有錯，
應該要用 `Buffer.byteLength(data)`

**Comment by cwc329** [comment, marked as answer]
我來演示一次我怎麼去 debug
不過有些地方是經驗，
可能不是那麼有脈絡。

首先你可以看到錯誤訊息的一些關鍵來自
`SyntaxError: Unexpected token S in JSON as position 0`
`JSON.parse`
`hw2.js:66:34`
這邊我從我的經驗會知道：「JSON.parse 帶進去的 string 是 S 開頭，所以他在解析的時候出錯。」
那這個時候我就會去想要怎麼把這個字串弄出來
去 google 一下怎麼把 Buffer 轉成 string 之後
然後把 chunks 轉成字串去看一下發現如下
> SyntaxError: Unexpected end of JSON input
>  at JSON.parse ...

這邊就是 server 的回應，
看起來是我們傳過去的東西有問題。

可是如果在本地 `console.log(JSON.parse(data))` 都是正常的，
這個就牽扯到編碼的問題，
網路在傳送資料的時候，
是不會直接傳送中文字的，
而是會先經過編碼，
然後把編碼過後的資料傳出去。
中文字在經過編碼後的 length 不會是 local 用 JS 看到的 length，
我剛剛測試了一下，
一個中文字的 byteLength 會是 3，
而英數字是 1，
也就是說一個中文字會讓兩者相差 2。

再回去找，
可以找到在發送 request 設定 header 的時候，
```js
'Content-Length': data.length
```
也就是這段有錯，
因為使用的是比較短的 length，
所以資料在 server 那邊被解析的時候尾巴被剪掉，
導致不是一個 valid JSON string，
當把 Content-Length 設定成正確的值，
這時 server 就可以正確的解讀，
也就可以正確地執行然後新增。

至於 Content-Length 是怎麼影響的，
是影響傳過去的資料長度，
還是 Server 解析的資料長度，
這邊我就不是很清楚了，
可以請其他大神來講解一下。

**Reply by torai55** [reply]
>> SyntaxError: Unexpected end of JSON input
>> at JSON.parse ...
>
> **這邊就是 server 的回應，
> 看起來是我們傳過去的東西有問題。**

**是 server 的回應ㄛ**
這邊要 highlight 起來，有個笨蛋一直以為這是 local 端出錯的錯誤訊息。
那個笨蛋就是我😢
非常感謝助教給我亂問一通。

**Reply by allenliaooooooo** [reply]
原來如此！！我也在這一步卡住很久
後來是參考上期學長姐作業用 `querystring` 來帶中文（雖然不太懂不過總算成功了）
原來 Buffer 轉成 String 的問題出在 Content-Length，真的是學到了

**Reply by aszx87410** [reply]
content length 其實不用帶也可以，那個會自動幫你處理

**Reply by allenliaooooooo** [reply]
蛤！竟然是連帶都不用帶！
Node.js 的文件有帶，我就從來沒想過不要帶
竟然是陷阱
